src/OrleansRuntime/Catalog/Catalog.cs

```diff
@@ -716,17 +720,18 @@ private void CreateGrainInstance(string grainTypeName, ActivationData data, stri
             {
                 Grain grain;
 
+                data.ServiceScope = this.serviceScopeFactory.CreateScope();
                 //Create a new instance of the given grain type
-                grain = grainCreator.CreateGrainInstance(grainType, data.Identity);
+                grain = grainCreator.CreateGrainInstance(data.ServiceProvider, grainType, data.Identity);
 
                 //for stateful grains, install storage bridge
                 if (grain is IStatefulGrain)
                 {
                     SetupStorageProvider(grainType, data);
 
                     var storage = new GrainStateStorageBridge(grainType.FullName, data.StorageProvider);
-
-                    grain = grainCreator.CreateGrainInstance(grainType, data.Identity, stateObjectType, storage);
+                    // TODO: avoid double creation!
+                    grain = grainCreator.CreateGrainInstance(data.ServiceProvider, grainType, data.Identity, stateObjectType, storage);
```

Comment:
Could we fix this TODO in this PR? it seems like `grain` created above is used only for the type check. Could we not replace that check with `typeof(IStatefulGrain).IsAssignableFrom(grainType)`?